### PR TITLE
Proto-CodeExpr

### PR DESCRIPTION
**Pull request by balacij**
I think we're _almost_ there with CodeExpr. I have  few notes in the code that I will review in this PR shortly.

However, notably, in this PR, we split up some files, create a new proxy package `drasil-code-base`, and create CodeExpr as a one-to-one copy of Expr, removing New/Message/Field from the original Expr.

### CODE REVIEW

File: code/drasil-code-base/Language/Drasil/Code/Expr.hs
Comment (JacquesCarette) [inline code/drasil-code-base/Language/Drasil/Code/Expr.hs:10]: Maybe it would make sense to share all these `Op` types? They are not recursive, so they are independent.

File: code/drasil-code-base/Language/Drasil/Code/Expr.hs
Comment (JacquesCarette) [inline code/drasil-code-base/Language/Drasil/Code/Expr.hs:60]: This is the one that can't be shared.

File: code/drasil-code-base/Language/Drasil/Code/Expr/Precedence.hs
Comment (JacquesCarette) [inline code/drasil-code-base/Language/Drasil/Code/Expr/Precedence.hs:12]: And then a lot of these could be shared too.

File: code/drasil-code-base/Language/Drasil/Code/Expr/Render.hs
Comment (JacquesCarette) [inline code/drasil-code-base/Language/Drasil/Code/Expr/Render.hs:13]: `render` made sense when it was more of a printing operation, and this propagated.  A simple `toCode` might be better, but I'm open to suggestions that are less printing-oriented.

Comment (JacquesCarette) [reply code/drasil-code-base/Language/Drasil/Code/Expr/Render.hs:13]: Also, because this whole file is about 'rendering', there is no point embedding that in the name too.  So `expr`, etc, might end up being best?

File: code/drasil-code/Language/Drasil/Chunk/CodeDefinition.hs
Comment (balacij) [inline code/drasil-code/Language/Drasil/Chunk/CodeDefinition.hs:65]: I'm not yet sure if we want to change `ODEInfo`s to use `CodeExpr`s instead of `Expr`s.
 https://github.com/JacquesCarette/Drasil/blob/5232d494995ca2d558dc60c2c81a8d4682a3c2e9/code/drasil-code/Language/Drasil/Data/ODEInfo.hs#L14-L23

Any thoughts on this?

File: code/drasil-code/Language/Drasil/Code/ExtLibImport.hs
Comment (JacquesCarette) [inline code/drasil-code/Language/Drasil/Code/ExtLibImport.hs:7]: It surprises me that `getActorName` is exported by `Language.Drasil`. Wouldn't this be move to `drasil-code`?

File: code/drasil-code/Language/Drasil/Chunk/CodeDefinition.hs
Comment (balacij) [reply code/drasil-code/Language/Drasil/Chunk/CodeDefinition.hs:65]: I would like to change it to use CodeExprs, but I'm not quite sure how the ODEs would use any objects as of right now.

File: code/drasil-code/Language/Drasil/Code/Imperative/Helpers.hs
Comment (JacquesCarette) [inline code/drasil-code/Language/Drasil/Code/Imperative/Helpers.hs:18]: Can you open an issue to get rid of this function? We shouldn't doing 'introspection' like this.

File: code/drasil-example/Drasil/GlassBR/ModuleDefs.hs
Comment (JacquesCarette) [inline code/drasil-example/Drasil/GlassBR/ModuleDefs.hs:91]: This is very interesting. `slope` is not necessarily a `CodeExpr` generator (though that is indeed how it is used). It kind of wants to be polymorphic over both [though I don't want to go there right now.]

Other routines below really are "pure code".

I don't want you to change anything here, this is just a comment.

File: code/drasil-lang/Language/Drasil/Constraint.hs
Comment (JacquesCarette) [inline code/drasil-lang/Language/Drasil/Constraint.hs:17]: This is a nice generalization.

**Review by JacquesCarette — CHANGES_REQUESTED** [review]

File: code/drasil-code/Language/Drasil/Chunk/CodeDefinition.hs
Comment (JacquesCarette) [reply code/drasil-code/Language/Drasil/Chunk/CodeDefinition.hs:65]: I think all of this can (and should) be changed to `CodeExpr`. If we need an ODE representation as an `Expr`, it will have to have its own definition there.

Comment (balacij) [reply code/drasil-code/Language/Drasil/Chunk/CodeDefinition.hs:65]: Sounds good. I'll get this done and let you know when it's completed.

File: code/drasil-code/Language/Drasil/Code/Imperative/Helpers.hs
Comment (balacij) [reply code/drasil-code/Language/Drasil/Code/Imperative/Helpers.hs:18]: Ah, yes, we had this issue before actually. #2414 is the related ticket, we only have 1 usage of it -- and it's very small and in `drasil-code`, I think we can add it to the scope of this PR too.

File: code/drasil-example/Drasil/GlassBR/ModuleDefs.hs
Comment (balacij) [reply code/drasil-example/Drasil/GlassBR/ModuleDefs.hs:91]: Thanks, yes, I agree. I believe this particular piece of code originally (pre-me starting the Expr split up) had used the Num typeclass to constrain the points so this code was more generalized then.

Comment (balacij) [reply code/drasil-example/Drasil/GlassBR/ModuleDefs.hs:91]: I'm not yet sure how this piece will look like when we have typed Expr, but hopefully it will revert to the original Num constraint (maybe it was Num, Fractional actually).

File: code/drasil-lang/Language/Drasil/Constraint.hs
Comment (balacij) [reply code/drasil-lang/Language/Drasil/Constraint.hs:17]: Thanks!

File: code/drasil-code-base/Language/Drasil/Code/Expr/Render.hs
Comment (balacij) [reply code/drasil-code-base/Language/Drasil/Code/Expr/Render.hs]: Yep, I agree. I changed the name of the function to just 'expr', and I also renamed the file to Convert.hs, I think it's a bit more suitable.

File: code/drasil-code/Language/Drasil/Chunk/CodeDefinition.hs
Comment (balacij) [reply code/drasil-code/Language/Drasil/Chunk/CodeDefinition.hs]: Fixed -- I think it looks a lot better now. Thanks!

File: code/drasil-code/Language/Drasil/Code/ExtLibImport.hs
Comment (balacij) [reply code/drasil-code/Language/Drasil/Code/ExtLibImport.hs:7]: `getActorName` is another function like `getInnerSpace` -- exact same functionality but instead of it being for the "Vect" Space, it's for the "Actor" Space. I guess, if anything, it would be another function marked for removal, like `getInnerSpace`.

**Review by JacquesCarette — APPROVED** [review]

File: code/drasil-code-base/Language/Drasil/Code/Expr.hs
Comment (balacij) [reply code/drasil-code-base/Language/Drasil/Code/Expr.hs:10]: Would we ever want to add some sort of typing to `CodeExpr`? If we do, we might want to keep it as is and continue to have it closely follow `Expr` in development, otherwise, I think we should definitely share the `Op` types, it would definitely tidy things up!

Comment (JacquesCarette) [reply code/drasil-code-base/Language/Drasil/Code/Expr.hs:10]: We will want to have both `Expr` and `CodeExpr` be typed. So if sharing might interfere with doing that, let's wait.

Comment (balacij) [reply code/drasil-code-base/Language/Drasil/Code/Expr.hs:10]: Sounds good, let's wait a bit to see how it ends up. Thanks!

File: code/drasil-code/Language/Drasil/Code/Imperative/Helpers.hs
Comment (balacij) [reply code/drasil-code/Language/Drasil/Code/Imperative/Helpers.hs:18]: There's a few too many to add on, I actually confused `getUpperBound` above with another, `getInnerSpace` which I used in an example. Let's leave it to another PR -- I'll try to get it done today however.

File: code/drasil-example/Drasil/NoPCM/ODEs.hs
Comment (balacij) [inline code/drasil-example/Drasil/NoPCM/ODEs.hs:19]: I forgot to mention this -- I moved the ODEs to a new file so that we could create `CodeExpr`s instead of `Expr`s by importing `Language.Drasil.CodeExpr` instead of `Language.Drasil` -- it has the same function names and operators, so it looks the same but creates different things (CodeExpr vs Expr). If we wanted to keep it in the original Body.hs file, we'd probably need to import the CodeExpr with a qualified name, which would also make the equations a bit more difficult to read.

**Comment by balacij** [comment]
Contributes to #2602

File: code/drasil-example/Drasil/NoPCM/ODEs.hs
Comment (JacquesCarette) [reply code/drasil-example/Drasil/NoPCM/ODEs.hs:19]: Saw that - good move.

**Review by JacquesCarette — APPROVED** [review]

**Comment by JacquesCarette** [comment]
I'm actually going to hold off on merging, and let the one from @Awurama-N come first. There will likely be non-trivial merge conflicts, and I'll want you to deal with them.

**Comment by balacij** [comment]
Sounds good. Thank you! :smile:

**Pull request merged**
